.github/workflows/build_and_publish.yml

```diff
@@ -0,0 +1,37 @@
+name: Build and upload to PyPI
+
+on:
+  pull_request:
+  release:
+    types:
+      - published
+
+jobs:
+  build_sdist:
+    name: Build source distribution
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Build sdist
+        run: pipx run build --sdist
+
+      - uses: actions/upload-artifact@v3
+        with:
+          path: dist/*.tar.gz
+
+  upload_pypi:
+    name: Upload to PyPI
+    needs: [build_sdist]
+    runs-on: ubuntu-latest
+    if: github.event_name == 'release' && github.event.action == 'published'
+    steps:
+      - uses: actions/download-artifact@v3
+        with:
+          name: artifact
+          path: dist
+
+      - uses: pypa/gh-action-pypi-publish@v1.5.0
+        with:
+          user: __token__
+          password: ${{ secrets.PYPI_TOKEN }}
```

Comment:
Yes we do. Each job creates a new environment, so these files are not automatically shared.

Reply:
While relying on other actions less would be nice, it makes sense to keep them because of other upcoming changes. 

This workflow will include a parametrised job that will build wheels for each OS in parallel. We want to do a bulk upload after all those builds succeed, therefore keeping upload in a separate job is necessary.